composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+        if todo_shards:
+            thread = Thread(target=self._download_thread, args=(todo_shards, part_min_id, part_max_id), daemon=True)
+            thread.start()
+        else:
+            self._done_loading()
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            int: Dataset length.
+        """
+        return math.ceil(self.index.total_samples / dist.get_world_size())
+
+    def _unpack_sample(self, data: bytes) -> Dict[str, Any]:
+        """Unpack a sample dict from raw bytes.
+
+        First unpacks the str to raw bytes dict, then unpacks each field's raw bytes.
+
+        Args:
+            data (bytes): The packed bytes of the sample.
+
+        Returns:
+            Dict[str, Any]: The sample dict.
+        """
+        key_to_raw = bytes_to_sample_dict(data, self.index.fields)
+        obj = {}
+        for key, decode in self.decoders.items():
+            value = key_to_raw[key]
+            if decode:
+                value = decode(value)
+            obj[key] = value
+        return obj
+
+    def __getitem__(self, idx: int) -> Any:
+        """Get the sample at the index, assuming its shard is loaded.
+
+        Do not call this directly unless all shards have been loaded. Will crash
+        if the shard is not loaded.
+
+        Args:
+            idx (int): Sample ID.
+
+        Returns:
+            Any: The sample.
+        """
+        shard = self.index.sample_shards[idx]
+        offset = self.index.sample_shard_offsets[idx]
+        size = self.index.bytes_per_sample[idx]
+
+        basename = get_shard_basename(shard)
+        shard_filename = os.path.join(self.local, basename)
+        fp = open(shard_filename, 'rb', 0)
+        fp.seek(offset)
+        data = fp.read(size)
+        fp.close()
+
+        return self._unpack_sample(data)
+
+    def _new_growing_epoch(self) -> int:
+        """Start a new growing epoch, in which we own the sample sequence because it grows.
+
+        Returns:
+            int: The epoch ID, an identifier which is given back to the caller.
+        """
+        with self._lock:
+            epoch = self._next_epoch
+            self._next_epoch += 1
+            self._epoch_to_todo_ids[epoch] = list(self._downloaded_ids)
+        return epoch
+
+    def _next_id(self, epoch: int) -> Optional[int]:
+        """Get next sample of the growing epoch given by epoch, or None if done.
+
+        If we are currently out of samples but not finished downloading the
+        shards, blocks until it has new samples.
+
+        Args:
+            epoch (int): The epoch, an identifier for this sequence of samples.
+
+        Returns:
+            int: ID of next sample.
+        """
+        while True:
+            with self._lock:
+                todo_ids = self._epoch_to_todo_ids[epoch]
+                if todo_ids:
+                    return todo_ids.pop(0)
```

Comment:
Can we pop a random element if shuffling (rather than always the 0th)? (If popping a random element, can we do it in `O(1)` by moving the last element into the position of the popped element?)

In addition, if `shuffle=False`, could we use a deque instead of a list (to avoid the `O(n^2)`)?

Reply:
I'll make a quick attempt at improving the runtime while still using a fully random permutation, but if it ends up being hard to read I'm going to defer to a future PR.

Luckily there's something of a 2-case analysis here:
* if the data is being consumed from the iterator quickly, then `todo_ids` is short, and shuffling is fast
* if the data is being consumed from the iterator slowly, then `todo_ids` is long, and shuffling the ids is slow, but probably not an issue since the data is being consumed slowly anyways

Reply:
Going to postpone this as V2 of StreamingDataset with global shuffle will refactor this.